Write state also on data nodes if not master eligible

brwe opened the pull request:
When a node was a data node only then the index state was not written.
In case this node connected to a master that did not have the index
in the cluster state, for example because a master was restarted and
the data folder was lost, then the indices were not imported as dangling
but instead deleted.
This commit makes sure that index state for data nodes is also written
if they have at least one shard of this index allocated.

I am a little lost with this. I found that the index can still be deleted
from a data node if the state was written but the node gets a new cluster state from a 
master that does not have it, for example because it was restarted without data folder. Happens
if the data node does not get the initial cluster state from the new but a later one and state
persistence is not disabled. 
I avoid this now by this: https://github.com/elasticsearch/elasticsearch/pull/9952/files#diff-f0f71bedb3d7e6f1cec54e8dddf5c3d3R109
but am worried about side effects this might have.  Any feedback appreciated.

closes #8823

kimchy commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -219,6 +217,22 @@ public void clusterChanged(ClusterChangedEvent event) {
+                        // remove the index state for this index if it is only a data node
+                        // only delete if the last shard was removed
+                        if (shardsAllocatedOnThisNodeInLastClusterState) {
+                            removeIndexState(indexMetaData);
```
I am not sure we need this removal logic, we already have that in `IndicesClusterStateService#applyCleanedIndices`, where an index is removed is no shards are around any more, its nice that it would be in a single place?

kimchy replied:
ohh, I think I see what happens, in IndicesClusterStateService, we just remove the index and not delete it from the file system, which is good. In IndicesStore, only there we actually delete the shard content once the shard is allocated on all the other nodes. I think that we need to add a logic there that if the there are no more shards around, we need to also delete the index itself (which will delete its metadata).

kimchy commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -219,6 +217,22 @@ public void clusterChanged(ClusterChangedEvent event) {
             for (IndexMetaData indexMetaData : newMetaData) {
+
+                boolean shardsAllocatedOnThisNodeInLastClusterState = true;
+                if (isDataOnlyNode(state)) {
```
I think this logic can be simplified a lot if we don't need to deal with deletes? That continue block at the end is confusing to me?

s1monw replied:
I don't think we need to check this everytime. we can just assign a boolean in the ctor if we need to write the state of not and if we need to write all states or not and centralize the decision. Then we can also remove the `isDataOnlyNode` method

s1monw commented on src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java:
```diff
@@ -190,14 +190,14 @@ protected Directory newDirectory(Path dir) throws IOException {
                 return Files.isRegularFile(entry)
                         && entryFileName.startsWith(prefix) // only state files
-                        && currentStateFile.equals(entryFileName) == false; // keep the current state file around
+                        && entryFileName.equals(currentStateFile) == false; // keep the current state file around
```
why is this change needed?

s1monw commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -219,6 +217,22 @@ public void clusterChanged(ClusterChangedEvent event) {
+
+                boolean shardsAllocatedOnThisNodeInLastClusterState = true;
+                if (isDataOnlyNode(state)) {
+                    boolean shardsCurrentlyAllocatedOnThisNode = shardsAllocatedOnLocalNode(state, indexMetaData);
```
can we replace this entire block with just this:
```
 if (indicesService.hasIndex(indexMetaData.index())) {
                        
}
```

and drop the entire delete logic here I think we can just fix the writing here. If we don't wipe the stuff correctly that's a different issue.

s1monw commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -344,6 +361,43 @@ public void onFailure(Throwable e) {
+        return ((state.nodes().localNode().masterNode() == false) && (state.nodes().localNode().dataNode() == true));
+    }
+
+    private void removeIndexState(IndexMetaData indexMetaData) {
```
this method must go away. We do index cleaning in IndicesService once indices are deleted. If it's buggy please open a different issue

kimchy replied:
@s1monw see my relevant comment for it above, I don't think we need this removal logic at all in this class? `IndicesClusterStateService#applyCleanedIndices` already removes an index is no shards are allocated on a node.

brwe replied:
I think it does not currently, only removes shards, the folder and meta data remain even if shards are allocated away. But I can change it to do this, would indeed simplify the code.

s1monw replied:
yeah lets remove it and open a new issue if needed.

s1monw commented on src/test/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -569,7 +569,12 @@ private NodeAndClient buildNode(int nodeId, long seed, Settings settings, Versio
         settings = getSettings(nodeId, seed, settings);
-        String name = buildNodeName(nodeId);
+        String name;
+        if (settings.get("node.name") != null) {
```
can we prevent this. I think we have to lock this naming thing down here and it should be owned by the test cluster

brwe replied:
It does make it easy to read the test and the logs if we have full control over naming. It is not often needed I guess but for the tests above I found it incredible helpful. What is the harm when we allow to provide custom names?

s1monw replied:
yes I really want this to be predictable and deterministic. we have to prevent arbitrary naming here IMO

s1monw commented on src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java:
```diff
@@ -165,4 +169,16 @@ public boolean nodesAdded() {
         return nodesRemoved() || nodesAdded();
     }
+
+    public boolean newMaster() {
```
can this be a private method?

s1monw commented on src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java:
```diff
@@ -105,6 +106,9 @@ public boolean indexRoutingTableChanged(String index) {
+        if (newMaster()) {
+            return ImmutableList.of();
+        }
         if (previousState == null) {
```
can this be `if (previousState == null || newMaster())`?

s1monw commented:
I left some comments...

brwe commented:
Made a pr for the deletion of index folders here: #9985 Should be easy to remove all the additional deletion code from this pr.

brwe commented:
#9985  is merged, I rebased on latest master and changed the code accordingly. I wanted to remove the change in ClusterStateEvent also because I was unable to reproduce the failures I'd seen before without it. But now I found that without the change the tests only pass on my Linux machine but fail every 10 iterations or so on my mac so something is still fishy. I'll try to come up with a detailed failure analysis tomorrow.

brwe commented:
I think I know what is going on now: The fresh master with the empty cluster state does (rarely) not send the first cluster state due to a race condition in lifecycles of DiscoveryService and its member Discovery. In DiscoveryService.doStart() the Discovery is started but the lifecyle for DiscoveryService is started only after that. This is why when the first cluster state reaches DiscoveryService.publish the lifecycle might or might not have started.

I added a commit d69f2cf1ac778cc0cfea37b351ae6ff5f1039262 where I removed the ClusterStateEvent change and added an artificial delay to the DiscoveryService.doStart() so that the tests fail reliably just so you can check if you want.

I would suggest we remove the ClusterStateEvent workaround and open another issue for this because this behavior is not a result of this pull request.

s1monw commented:
I agree with your idea of opening a new issue for the ClusterSTateEvent problem

brwe commented:
Chatted with @s1monw and now rewrote it so that the selection of what to write is not done in GatewayMetaState anymore. I tried to do it similar to #10016. It is still a little raw but but would be great if you could let me know if this is the right direction.

s1monw commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -90,6 +94,26 @@ public MetaData loadMetaState() throws Exception {
     }
 
+    public static class IndexMetaWriteInfo {
+        IndexMetaData newMetaData;
```
can they all be package private and final please? just like a struct

s1monw replied:
also you might wanna put this to the end of the file

s1monw commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -114,33 +138,17 @@ public void clusterChanged(ClusterChangedEvent event) {
-                    continue;
-                }
+            Iterable<IndexMetaWriteInfo> writeInfo = new ArrayList<>();
+            if (isDataOnlyNode(event.state())) {
```
can we maybe do this like this:

```Java
final Iterable<IndexMetaWriteInfo> writeInfo;
if (a) {

} else if (b) {

} else {
  writeInfo = Collections.emptyList();
}
```

s1monw commented on src/main/java/org/elasticsearch/gateway/IndexMetaState.java:
```diff
@@ -0,0 +1,126 @@
+            } catch (IOException e) {
+                logger.debug("failed to load index state ", e);
+            }
+        } return currentIndexMetaData;
```
extra newline here after `}`

s1monw commented on src/main/java/org/elasticsearch/gateway/IndexMetaState.java:
```diff
@@ -0,0 +1,126 @@
+            String writeReason = null;
+            IndexMetaData currentIndexMetaData = maybeLoadIndexState(currentMetaData, indexMetaData);
+            // ... new index or state persistence was disabled?
+            if ( currentIndexMetaData == null) {
```
extra space after `(`

s1monw commented on src/main/java/org/elasticsearch/gateway/IndexMetaState.java:
```diff
@@ -0,0 +1,126 @@
+            } else if (currentIndexMetaData.version() != indexMetaData.version()) {
+                writeReason = "version changed from [" + currentIndexMetaData.version() + "] to [" + indexMetaData.version() + "]";
+            }
+            if (writeReason!=null) {
```
spaces needed before and after `!=`

s1monw commented on src/main/java/org/elasticsearch/gateway/IndexMetaState.java:
```diff
@@ -0,0 +1,126 @@
+    public Iterable<GatewayMetaState.IndexMetaWriteInfo> getIndicesToWriteDataOnlyNode(ClusterChangedEvent event, MetaData currentMetaData) {
+        Map<String, GatewayMetaState.IndexMetaWriteInfo> indicesToWrite = new HashMap<>();
+        RoutingNode thisNode = event.state().getRoutingNodes().node(event.state().nodes().localNodeId());
+        if ( thisNode == null) {
```
space after `(`?

s1monw commented on src/main/java/org/elasticsearch/gateway/IndexMetaState.java:
```diff
@@ -0,0 +1,126 @@
+        this.metaStateService = metaStateService;
+    }
+
+    public Iterable<GatewayMetaState.IndexMetaWriteInfo> getIndicesToWriteDataOnlyNode(ClusterChangedEvent event, MetaData currentMetaData) {
```
can we have some javadoc here? also the name is pretty long maybe we can rename to something like `filterStatesOnMaster` / `filterStateOnDataNode`

s1monw commented on src/main/java/org/elasticsearch/gateway/IndexMetaState.java:
```diff
@@ -0,0 +1,126 @@
+ * their state written on disk, but don't exists in the metadata of the cluster), and importing
+ * them into the cluster.
+ */
+public class IndexMetaState extends AbstractComponent {
```
I think this file can just be merged into  `GatewayMetaState` I don't think those two methods justify a new component?

brwe replied:
I thought it might be easier to test? I could write a single node test instead but this way it seemed easier to me. I have no strong feelings about it though so you decide.

s1monw commented on src/test/java/org/elasticsearch/gateway/GatewayMetaStateTests.java:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    @Test
+    public void testAllCombinations() throws Exception {
```
can we have a test for each of these names after what they are doing? that would be very helpful I think! I am ok with duplicating code here and there but these unittests should be more readable

s1monw commented:
I left a bunch of comments

brwe commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -114,33 +113,19 @@ public void clusterChanged(ClusterChangedEvent event) {
-                if (currentMetaData == null) {
-                    // a new event..., check from the state stored
-                    try {
-                        currentIndexMetaData = metaStateService.loadIndexState(indexMetaData.index());
```
Could we skip this loading from disk? I made the change so that we still load (see `maybeLoadIndexState()` below) if the meta data is not in memory but fail to understand why we do it. Is it an optimization?

s1monw replied:
as far as I am concerned we could just skip it since we write the new one anyway. Yet, we might have a new meta on disk but that would be a bug too no? so I think we can drop it?

brwe replied:
Ok, removed this check now.

brwe commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -225,9 +218,116 @@ private void pre20Upgrade() throws Exception {
+            // ... state persistence was disabled or index was newly created
+            if (currentIndexMetaData == null) {
+                writeReason = "freshly created";
+                // ... new shard is allocated on node (we could optimize here and make sure only written once and not for each shard per index -> do later)
```
Not sure if we should check here if we wrote the state already.

s1monw replied:
I think if  you move the version check above this we only do it once?

brwe replied:
cannot do that. version is incremented when mapping changes or setting etc, not when relocated. We need to somehow check for initializing.

brwe replied:
see below. check now if shard was allocated on node in previous cluster state and if not do not write.

brwe commented:
Addressed all comments. I am unsure about two things, left comments above about it: We did check on disk if we have the state already before writing in case the in memory state is null. However, this should be a rare event so I am unsure if we need this optimization. In addition, I made it so that whenever a shard is initializing on a node we write the meta state of this index and perform no check if we wrote before already. Should I check instead?

s1monw commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -154,6 +139,14 @@ public void clusterChanged(ClusterChangedEvent event) {
+    }
+
+    protected boolean isMasterEligibleNode(ClusterState state) {
+        return state.nodes().localNode().masterNode() == true;
```
the `== true` is obsolete

s1monw commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -154,6 +139,14 @@ public void clusterChanged(ClusterChangedEvent event) {
     }
 
+    protected boolean isDataOnlyNode(ClusterState state) {
+        return ((isMasterEligibleNode(state) == false) && (state.nodes().localNode().dataNode() == true));
```
the `== true` is obsolete

s1monw commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -225,9 +218,116 @@ private void pre20Upgrade() throws Exception {
+            try {
+                currentIndexMetaData = metaStateService.loadIndexState(indexMetaData.index());
+            } catch (IOException e) {
+                logger.debug("failed to load index state ", e);
```
can we put a TODO here to maybe bubble this Exception up in the future it seems risky to just swallow it....

brwe replied:
if I remove loading the state this method will go away anyway

s1monw commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -225,9 +218,116 @@ private void pre20Upgrade() throws Exception {
+     * @param currentMetaData the current index state in memory.
+     * @return iterable over all indices states that should be written to disk
+     */
+    public static Iterable<GatewayMetaState.IndexMetaWriteInfo> filterStateOnDataNode(ClusterChangedEvent event, MetaData currentMetaData, MetaStateService metaStateService, ESLogger logger) {
```
since we only use the `event().state()` can we just pass the state to it instead?

s1monw commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -225,9 +218,116 @@ private void pre20Upgrade() throws Exception {
+     * @param currentMetaData the current index state in memory.
+     * @return iterable over all indices states that should be written to disk
+     */
+    public static Iterable<GatewayMetaState.IndexMetaWriteInfo> filterStatesOnMaster(ClusterChangedEvent event, MetaData currentMetaData, MetaStateService metaStateService, ESLogger logger) {
```
since we only use the `event().state()` can we just pass the state to it instead?

s1monw commented:
I like this  a lot - left some minor comments

brwe commented:
Addressed all comments except for the version check  thing above. We cannot use index version to ensure that the state is written because the version is not necessarily updated for the index metadata if shards are relocated. But shards should initialize too often on a node so it might be ok if we write the same state each time we see an initializing shard?

brwe commented:
Chatted with @s1monw who reminded me that shards can be in initializing state for a while and while they are we would always write, so that is not a good idea. Will now instead check in the event if the last state had shards allocated and if not then write if the new one has.

brwe commented:
pushed a new commit to address this. we now check if a shard was already present in last cluster state and if so not write.

s1monw commented:
it LGTM I think @bleskes should take one more look

bleskes commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -72,7 +71,7 @@ public GatewayMetaState(Settings settings, NodeEnvironment nodeEnv, MetaStateSer
             nodeEnv.ensureAtomicMoveSupported();
         }
-        if (DiscoveryNode.masterNode(settings)) {
+        if (DiscoveryNode.masterNode(settings) || DiscoveryNode.dataNode(settings)) {
```
Wondering until when do we need the BWC bellow? (not saying remove now)

brwe replied:
The pre20Upgrade() we should still do because this change is going to 1.x as well. But we probably need not check on data nodes for pre 019 state. Would it harm to keep this check or should I explicitly only check  pre 019 on master nodes and pre 20 on master and data nodes?

 /**
     * Elasticsearch 2.0 deprecated custom routing hash functions. So what we do here is that for old indices, we
     * move this old & deprecated node setting to an index setting so that we can keep things backward compatible.
     */

brwe replied:
The pre20Upgrade() we should still do because this change is going to 1.x as well. But we probably need not check on data nodes for pre 019 state. Would it harm to keep this check or should I explicitly only check  pre 019 on master nodes and pre 20 on master and data nodes?

bleskes commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -225,9 +218,110 @@ private void pre20Upgrade() throws Exception {
+        RoutingNode newRoutingNode = newState.getRoutingNodes().node(newState.nodes().localNodeId());
+        RoutingNode previousRoutingNode = previousState.getRoutingNodes().node(newState.nodes().localNodeId());
+        if (newRoutingNode == null) {
+            // this needs some other handling
```
I think this is correct and we need an exception here. It means the routing node is not consistent or we have gotten here while the current node is not part of the CS, which is also broken.

bleskes commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -225,9 +218,110 @@ private void pre20Upgrade() throws Exception {
+            // this needs some other handling
+            return indicesToWrite.values();
+        }
+        // iterate over all shards allocated on this node in the new cluster state but only write if ...
```
only write if ... , what?

bleskes replied:
Oh. I see. Nevermind :)

bleskes commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -225,9 +218,110 @@ private void pre20Upgrade() throws Exception {
+
+        Map<String, GatewayMetaState.IndexMetaWriteInfo> indicesToWrite = new HashMap<>();
+        RoutingNode newRoutingNode = newState.getRoutingNodes().node(newState.nodes().localNodeId());
+        RoutingNode previousRoutingNode = previousState.getRoutingNodes().node(newState.nodes().localNodeId());
```
this one should go away. I see why it's here (because we only cache the currentMetaData which doesn't tell us anything about allocation) but now we have too ways of deciding what to do . One based on last written state and the other based on the last CS in the event. This is an issue. Can we cache a set of last written indices next to `currentMetaData` and use that?

bleskes commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -225,9 +218,110 @@ private void pre20Upgrade() throws Exception {
+    }
+
+    public static boolean shardsOfThisIndexExistedOnNodeAlready(RoutingNode previousRoutingNode, MutableShardRouting shardRouting) {
+        return previousRoutingNode.shardsWithState(shardRouting.index(), ShardRoutingState.INITIALIZING, ShardRoutingState.STARTED).size() != 0;
```
this is tricky as it misses `ShardRoutingState#RELOCATING`. Can we add a method that just returns all the shards of an index? or alternatively, if we follow  my previously suggestion of caching written indices - just check the new index if it is not part of the written indices set.

bleskes commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -225,9 +218,110 @@ private void pre20Upgrade() throws Exception {
+        Map<String, GatewayMetaState.IndexMetaWriteInfo> indicesToWrite = new HashMap<>();
+        MetaData newMetaData = state.metaData();
+        // iterate over all indices but only write if ...
+        for (IndexMetaData indexMetaData : newMetaData) {
```
I think we can merge this method body with the different one, no? the only difference is which indices we begin with (all of them or just those allocated to this node).

bleskes commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -114,33 +113,19 @@ public void clusterChanged(ClusterChangedEvent event) {
-                }
+            Iterable<IndexMetaWriteInfo> writeInfo;
+            if (isDataOnlyNode(event.state())) {
+                writeInfo = filterStateOnDataNode(event, currentMetaData);
```
I know this was already there, but can we rename currentMetaData to previouslyWrittenMetaData ? I think it will be easy to understand what's what here..

brwe replied:
Agree, the name is confusing. But the meta data is not always written, so maybe call it previousMetaData?

bleskes commented:
Looks good in general. Left some comments that I think will simplify things.

brwe commented:
Implemented all suggestions. 
@bleskes thanks for the tip with the indices list caching, it simplifies think indeed! Please take another look.

bleskes commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -57,7 +57,9 @@
-    private volatile MetaData currentMetaData;
+    private volatile MetaData previousMetaData;
+
+    private final Set<String> previouslyWrittenIndices = Collections.synchronizedSet(new HashSet<String>());
```
can we change this to volatile and use an ImmutableSet when we assign it (the members shouldn't be changed)

bleskes commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -150,10 +155,29 @@ public void clusterChanged(ClusterChangedEvent event) {
         }
     }
 
+    public static Set<String> getRelevantIndices(ClusterState state, Set<String> previouslyWrittenIndices) {
```
can we add some java docs about what relevant indices mean?

bleskes commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -225,9 +249,87 @@ private void pre20Upgrade() throws Exception {
+            indices.add(routing.index());
+        }
+        // we have to check the meta data also: closed indices will not appear in the routing table, but we must still write the state if we have it written on disk previously
+        for (ObjectCursor<String> index : state.metaData().indices().keys()) {
```
we can iterate on the state.metaData() it self. will make this loop slightly simpler

bleskes commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -225,9 +249,87 @@ private void pre20Upgrade() throws Exception {
+        Set<String> relevantIndices;
+        relevantIndices = new HashSet<>();
+        // we have to iterate over the metadata to make sure we also capture closed indices
+        for (ObjectCursor<String> index : state.metaData().indices().keys()) {
```
same here no need to use the keys()

bleskes commented on src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -225,9 +249,87 @@ private void pre20Upgrade() throws Exception {
+     * @param newMetaData                 The new metadata
+     * @return iterable over all indices states that should be written to disk
+     */
+    public static Iterable<GatewayMetaState.IndexMetaWriteInfo> filterStates(Set<String> previouslyWrittenIndices, Set<String> potentiallyUnwrittenIndices, MetaData previousMetaData, MetaData newMetaData) {
```
can we call this resolveStatesToBeWritten?

bleskes commented:
Left some very minor comments. Feel free to push without another review. LGTM!

brwe commented:
need to investigate https://github.com/elastic/elasticsearch/issues/10017 before we can push

brwe commented:
The reason why the tests failed on CI is the same I described in the beginning https://github.com/elastic/elasticsearch/pull/9952#issue-59520914 : a data node receives a new cluster state from a master that does not have the index in its state but the data node missed the state with a no master block before and so state persistence was not disabled. the fact that an index is not in the cluster state is then interpreted as delete command. This can happen here for the reasons described in #10017 but there might be other reasons as well. I now think we should not delete indices at all if the cluster state that would cause a deletion comes from a new master.
I added a new commit for this but need someone to confirm that this is actually the right solution.

s1monw commented:
> I added a new commit for this but need someone to confirm that this is actually the right solution.

+1 to the solution

brwe commented:
Chatted with @kimchy and we decided to push as is and add a //norelease comment and open an issue because the short term fix for the  problem (https://github.com/elastic/elasticsearch/pull/9952#issuecomment-97707128) is not very elegant.
Added another commit to address the latest comments.

s1monw commented:
LGTM
